### ISSUE TITLE
[RFC] Validate that FSDP modules execution order is the same across ranks

### ISSUE THREAD

**Issue by zhaojuanmao** [pytorch/pytorch#75373]
### 🚀 The feature, motivation and pitch

Right now FSDP assumes module execution order is the same across ranks without validation. This is not safe, it would either result in hang or incorrect results. 

Propose to add validation in the following way, alternative ways are welcomed. Note the following approach will work efficiently for graphs whose execution order do not change over iterations. If graph execution order changes after 1st iteration, it still can validate execution orders are the same across ranks, but that will be expensive, so we can chose to give users a warning like "graph execution orders are changed but they are not validated to be the same across ranks, it has risk to result in training hang or incorrect results".

The proposal assumes each rank wraps the same module in the same way:

1. During the first iteration, if module is the root, traverse all the children fsdp modules, and assign each fsdp module a id. e.g., fsdp_module1.id = traverse_index. 
2. In current fsdp implementation, root fsdp module keeps a list called "_fsdp_graph_order" to maintain the fsdp module references in the execution order. We can have a new list to put the active fsdp module's id "fsdp_module.id" into it, and trace the executed fsdp modules using the ids for both forward and backward pass . then at the end of backward pass, the new list will capture the execution order of fsdp modules for the first iteration. we can broadcast rank 0's execution order list to other ranks and let each rank check whether the execution order list matches or not. if not, throw errors. 
3. after 1st iteration, we can continue to trace the fsdp module execution orders using the ids, and check whether it is the same as the one captured in the first iteration. If the order changed after 1st iteration, give users a warning

**Comment by zhaojuanmao** [comment]
cc @awgu

**Comment by rohan-varma** [comment]
Thanks for this proposal it looks good to me overall especially if the concern below can be addressed:

Regarding (2): 

> then at the end of backward pass, the new list will capture the execution order of fsdp modules for the first iteration. we can broadcast rank 0's execution

If each rank creates its execution order independently during forward and backward pass, and then we broadcast it at the end, won't we risk running into hang in the first iteration itself?

For example, if rank 0 and rank 1 have mismatched execution order, we might call `all_gather` or `_reduce_scatter` on parameter/gradients of mismatching shapes, which possibly might result in a hang? Or are we relying on the asynchronous nature of CUDA and only doing sync at the end, which is why this broadcast and end of forward + backward would work?

Also, do we need to wait for the backwards pass to do this in general or would we get enough information from the forward pass?

Another option is to take an approach similar to TORCH_DISTRIBUTED_DEBUG=DETAIL which checks shapes of collectives before running them.

In the first iteration, during _rebuild_full_param for all_gather, could we just broadcast the all_gather tensor size and validate that is the same (for all FSDP units)? Since we're doing it only for the first iteration, presumably the perf hit shouldn't be too much. I believe this should capture most cases where there is a order execution mismatch, although this wouldn't capture the edge case where the out of order execution lines up exactly so that there are different layers used but they happen to have the same size.

**Comment by zhaojuanmao** [comment]
you are right, this proposal can not prevent hang in the first iteration. When there is hang in the first iteration, it will become the same problem as DDP: there should be some desync happened, in this case hang and timeout will crash the process, and "TORCH_DISTRIBUTED_DEBUG=DETAIL " can help debug this, right? 

but this proposal can help surface the accuracy issue when the executor order is not the same and there is no hang happened, in which case the accuracy issue will be hidden and very hard to debug.

**Comment by awgu** [comment]
Just curious, what are practical examples for which the module execution order may differ across ranks?

**Comment by awgu** [comment]
Is the backward pass order in terms of the start of a parameter's gradient computation or the end? It seems like start makes sense.

**Comment by zhaojuanmao** [comment]
> Just curious, what are practical examples for which the module execution order may differ across ranks?

so far has not heard users reported this mismatch orders. But I remember Tesla could train different sub graphs among ranks in their models, DDP supported this use case. Some complex ads model may train different sub graphs among ranks as well
